src/topology/perfect.lean

```diff
@@ -211,3 +218,342 @@ begin
+  { intros m hm,
+    rw nat.lt_succ_iff_lt_or_eq at hm,
+    rw [← reverse_inj] at h,
+    simp only [res_succ, concat_eq_append, reverse_append, reverse_singleton,
+      singleton_append, reverse_inj] at h,
+    cases hm with hm hm,
+    { exact ih h.2 _ hm },
+    rw hm,
+    exact h.1, },
+  { simp },
+  simp only [res_succ],
+  rw [ih _, h _ (nat.lt_succ_self _)],
+  intros m hmn,
+  exact h m (hmn.trans (nat.lt_succ_self _)),
+end
+
+/-- Two infinite sequences are equal if and only if all their restrictions are.-/
+theorem eq_iff_res_eq (x y : ℕ → α) : (∀ n, res x n = res y n) ↔ x = y :=
+begin
+  split; intro h,
+  { ext n,
+    specialize h n.succ,
+    rw res_eq_iff at h,
+    exact h _ (nat.lt_succ_self _), },
+  rw h,
+  simp,
+end
+
+/-- `cylinder x n` is equal to the set of sequences `y` with the same restriction to `n` as `x`.-/
+theorem cylinder_eq_res (x : ℕ → α) (n : ℕ) : pi_nat.cylinder x n = {y | res y n = res x n} :=
+begin
+  ext y,
+  dsimp[pi_nat.cylinder],
+  rw res_eq_iff,
+end
+
+/-- From a `β`-scheme on `α` `A`, we define a partial function from `(ℕ → β)` to `α`
+which sends each infinite sequence `x` to an element of the intersection along the
+branch corresponding to `x`, if it exists.
+We call this the map induced by the scheme. -/
+noncomputable def map : Σ s : set (ℕ → β), s → α :=
+⟨λ x, set.nonempty ⋂ n : ℕ, A (res x n), λ ⟨x, hx⟩, hx.some⟩
+
+/-- A scheme is antitone if each set contains its children.  -/
+def antitone : Prop := ∀ l : list β, ∀ a : β, A (l.concat a) ⊆ A l
+
+/-- A useful strengthening of being antitone is to require that each set contains
+the closure of each of its children. -/
+def closure_antitone [topological_space α] : Prop :=
+∀ l : list β, ∀ a : β, closure(A (l.concat a)) ⊆ A l
+
+/-- A scheme is disjoint if the children of each set of pairwise disjoint. -/
+def disjoint : Prop := ∀ l : list β, ∀ a b : β, a ≠ b →
+  disjoint (A (l.concat a)) (A (l.concat b))
+
+/-- A scheme on a metric space has vanishing diameter if diameter approaches 0 along each branch. -/
+def vanishing_diam [pseudo_metric_space α] : Prop :=
+∀ x : ℕ → β, tendsto (λ n : ℕ, emetric.diam (A (res x n))) at_top (𝓝 0)
+
+variable {A}
+
+/-- If `x` is in the domain of the induced map of a scheme `A`,
+its image under this map is in each set along the corresponding branch. -/
+lemma map_mem {x : ℕ → β} (hx : x ∈ A.map.1) (n : ℕ) : A.map.2 ⟨x, hx⟩ ∈ A (res x n) :=
+begin
+  have := hx.some_mem,
+  rw mem_Inter at this,
+  exact this n,
+end
+
+lemma antitone_of_closure_antitoine [topological_space α] (hA : closure_antitone A) : antitone A :=
+λ l a, subset_closure.trans (hA l a)
+
+lemma closure_antitone_of_antitone_of_is_closed [topological_space α] (hanti : antitone A)
+  (hclosed : ∀ l, is_closed (A l)) : closure_antitone A :=
+begin
+  intros l a,
+  rw (hclosed _).closure_eq,
+  apply hanti,
+end
+
+@[nolint ge_or_gt]
+lemma small_dist_of_vanishing_diam [pseudo_metric_space α] (hA : vanishing_diam A)
```

Comment:
```suggestion
lemma vanishing_diam.small_dist [pseudo_metric_space α] (hA : vanishing_diam A)
```
Possibly change it to `vanishing_diam.something_clearer`

Reply:
I went with vanishing_diam.dist_lt